nebula_decoders/include/nebula_decoders/nebula_decoders_velodyne/decoders/velodyne_generic_decoder.hpp

```diff
@@ -0,0 +1,445 @@
+    if (!scan_pc_->points.empty()) {
+      auto current_azimuth = scan_pc_->points.back().azimuth;
+      auto phase_diff = (2 * M_PI + current_azimuth - phase);
+      while (phase_diff < M_PI_2 && !scan_pc_->points.empty()) {
+        overflow_pc_->points.push_back(scan_pc_->points.back());
+        scan_pc_->points.pop_back();
+        current_azimuth = scan_pc_->points.back().azimuth;
+        phase_diff = (2 * M_PI + current_azimuth - phase);
+      }
+      overflow_pc_->width = overflow_pc_->points.size();
+      scan_pc_->width = scan_pc_->points.size();
+      scan_pc_->height = 1;
+    }
+    return std::make_tuple(scan_pc_, scan_timestamp_);
+  }
+
+  // DONE
+  int pointsPerPacket() { return BLOCKS_PER_PACKET * CHANNELS_PER_BLOCK; }
+
+  // DONE
+  void reset_pointcloud(size_t n_pts, double time_stamp)
+  {
+    //  scan_pc_.reset(new NebulaPointCloud);
+    scan_pc_->points.clear();
+    max_pts_ = n_pts * pointsPerPacket();
+    scan_pc_->points.reserve(max_pts_);
+    reset_overflow(time_stamp);  // transfer existing overflow points to the cleared pointcloud
+  }
+
+  // DONE
+  void reset_overflow(double time_stamp)
+  {
+    if (overflow_pc_->points.size() == 0) {
+      scan_timestamp_ = -1;
+      overflow_pc_->points.reserve(max_pts_);
+      return;
+    }
+
+    // Compute the absolute time stamp of the last point of the overflow pointcloud
+    const double last_overflow_time_stamp =
+      scan_timestamp_ + 1e-9 * overflow_pc_->points.back().time_stamp;
+
+    // Detect cases where there is an unacceptable time difference between the last overflow point
+    // and the first point of the next packet. In that case, there was probably a packet drop so it
+    // is better to ignore the overflow pointcloud
+    if (time_stamp - last_overflow_time_stamp > 0.05) {
+      scan_timestamp_ = -1;
+      overflow_pc_->points.clear();
+      overflow_pc_->points.reserve(max_pts_);
+      return;
+    }
+
+    // Add the overflow buffer points
+    while (overflow_pc_->points.size() > 0) {
+      auto overflow_point = overflow_pc_->points.back();
+
+      // The overflow points had the stamps from the previous pointcloud. These need to be changed
+      // to be relative to the overflow's packet timestamp
+      double new_timestamp_seconds =
+        scan_timestamp_ + 1e-9 * overflow_point.time_stamp - last_block_timestamp_;
+      overflow_point.time_stamp =
+        static_cast<uint32_t>(new_timestamp_seconds < 0.0 ? 0.0 : 1e9 * new_timestamp_seconds);
+
+      scan_pc_->points.emplace_back(overflow_point);
+      overflow_pc_->points.pop_back();
+    }
+
+    // When there is overflow, the timestamp becomes the overflow packets' one
+    scan_timestamp_ = last_block_timestamp_;
+    overflow_pc_->points.clear();
+    overflow_pc_->points.reserve(max_pts_);
+  }
+
+  void unpack(const velodyne_msgs::msg::VelodynePacket & velodyne_packet)
+  {
+    const raw_packet_t * raw = (const raw_packet_t *)&velodyne_packet.data[0];
```

Comment:
Never use C-style casts for converting byte arrays to structs. This is undefined behavior.
Use `memcpy` instead: please refer to [the Hesai decoder](https://github.com/tier4/nebula/blob/affb6196b91731f66efb7ecc1bcd1866ef3c2c99/nebula_decoders/include/nebula_decoders/nebula_decoders_hesai/decoders/hesai_decoder.hpp#L57C1-L73C4) for a safe implementation.

Please also move parsing to the existing `parsePacket` member function for readability.